### PR TITLE
Use macros to add parameters in PMP

### LINKED ISSUE #1798: Improve CGAL named parameters with macros
- Adding new named parameter in `Polygon_mesh_processing/internal/named_function_params.h` could be improved if a system with macros was used. Since there are two functions to add per named parameter, I think a mechanism like `Kernel_23/include/CGAL/Kernel/interface_macros.h`.
- The namespace CGAL is polluted, all classes should be in a nested namespace (`parameters`?)

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes
- declares two macro in named_fucntion_params.h that help adding new parameters
- replaces the parameters declarations by calls to the macro if possible

## Release Management

* Affected package(s):
* Issue(s) solved (if any): fix #1798 
* Feature/Small Feature (if any):

### CODE REVIEW

**Comment by sloriot** [comment]
This only address the first part of the issue.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/named_function_params.h [hunk 1]
```diff
@@ -28,42 +28,93 @@
...
-  enum number_of_points_on_faces_t  { number_of_points_on_faces };
-  enum number_of_points_per_face_t  { number_of_points_per_face };
-  enum grid_spacing_t               { grid_spacing };
-  enum nb_points_per_area_unit_t    { nb_points_per_area_unit };
-  enum number_of_points_per_edge_t  { number_of_points_per_edge };
-  enum number_of_points_on_edges_t  { number_of_points_on_edges };
-  enum nb_points_per_distance_unit_t{ nb_points_per_distance_unit };
-
+  enum all_default_t { all_default} ; //cannot use macro because of inline
+  enum nb_points_per_area_unit_t    { nb_points_per_area_unit }; //cannot use macro because names diverge
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/named_function_params.h]: You can add a third parameter to the macro to handle those cases

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/named_function_params.h [hunk 2]
```diff
@@ -28,42 +28,93 @@
...
+#define Cgal_add_pmp_parameter_in_struct(X_t, X)       \
+  template<typename K>                                 \
+  pmp_bgl_named_params<K, X_t, self>                   \
+  X(const K& k) const                                  \
+  {                                                    \
+    typedef pmp_bgl_named_params<K, X_t, self> Params; \
+    return Params(k, *this);                           \
+  }
+
+Cgal_add_pmp_parameter_and_enum(geom_traits_t, geom_traits)
```
Comment (maxGimeno) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/named_function_params.h]: What do you mean "the solution similar to `interface_macros.h` " ? A system that asks each file to define its own macro ?

**Comment by lrineau** [comment]
@sloriot It seems @maxGimeno has fixed the issues pointed by your review. Can you please make a new review of the PR now?

The Travis builds seem OK too.

**Comment by lrineau** [comment]
Build failure in Travis: https://travis-ci.org/CGAL/cgal/jobs/199999707

**Comment by lrineau** [comment]
@sloriot Do you approve recent changes?

File: BGL/include/CGAL/boost/graph/boost_parameters_interface.h [hunk 3]
```diff
@@ -19,6 +19,6 @@
 
 
 
-CGAL_add_pmp_parameter(vertex_index_t, vertex_index_map, vertex_index_map)
-CGAL_add_pmp_parameter(edge_index_t, edge_index_map, edge_index_map)
-CGAL_add_pmp_parameter(graph_visitor_t, visitor, visitor)
+CGAL_add_named_parameter(vertex_index_t, vertex_index_map, vertex_index_map)
```
Comment (lrineau) [inline BGL/include/CGAL/boost/graph/boost_parameters_interface.h]: While you are at renaming the macro, could you follow CGAL guidelines and use a full-uppercase name?

Comment (maxGimeno) [inline BGL/include/CGAL/boost/graph/boost_parameters_interface.h]: Actually the guidelines say : 
- Names of constants are uppercase (e.g., ORIGIN).
- Names of macros should begin with the prefix CGAL_.
 So I think it does match the naming convention. Furthermore, macros in interface_macros.h are named CGAL_Kernel_obj, etc. 
 I agree that I should make the A of Add uppercase, but following the guidelines does not lead to CGAL_ADD_PMP_PARAMETER.

Comment (lrineau) [inline BGL/include/CGAL/boost/graph/boost_parameters_interface.h]: ... You are right! :-\

**Comment by maxGimeno** [comment]
Is this PR going to be integrated soon ? I'd like to use it for #2005

**Pull request merged**